### DISCUSSION TITLE
Build master branch on Windows (MinGW/MSYS)

### DISCUSSION THREAD

**Discussion by Shionigami** [ccache/ccache#630] [Show and tell]
I am currently trying to build ccache master branch with my Windows environment, using MinGW.
So I tried to follow the official instructions, using cmake and then make, but it only generates a Visual Studio sln file, and I got a lot of cmake errors.
For info, I was able to build with Linux without any problem.

Then i found the repo from cristianadam and followed the instructions I found there, to build with Ninja with the corresponding branch:
https://github.com/cristianadam/ccache/actions/runs/24884029/workflow

Here I get better results, but I still have issues with a library which is not detected properly:
```
FAILED: ccache.exe
cmd.exe /C "cd . && C:\msys64\usr\bin\c++.exe -O3 -DNDEBUG   CMakeFiles/ccache.dir/src/main.cpp.obj  -o ccache.exe -Wl,--out-implib,libccache.dll.a -Wl,--major-image-version,0,--minor-image-version,0  libccache_lib.a  -lws2_32  -static  -lgcc  -lstdc++  -lwinpthread  -dynamic  zstd-build/lib/libzstd.a  libb2-build/liblibb2.a  libb2-build/liblibblake2b_ref.a  libb2-build/liblibblake2b_sse2.a  libb2-build/liblibblake2b_ssse3.a  libb2-build/liblibblake2b_sse41.a  libb2-build/liblibblake2s_avx.a  libb2-build/liblibblake2b_xop.a && cd ."
/usr/lib/gcc/x86_64-pc-msys/9.3.0/../../../../x86_64-pc-msys/bin/ld: cannot find -lwinpthread
collect2: error: ld returned 1 exit status
[114/114] Linking CXX executable unittest.exe
ninja: build stopped: subcommand failed.
```

To summarize what I did:
- Cloned the repo from cristianadam, ran the following commands:
`cmake -S . -B build_ninja -DCMAKE_BUILD_TYPE=Release -GNinja -DCMAKE_MAKE_PROGRAM="C:\Tools\ninja\ninja.exe"
cmake --build build_ninja`
- I also installed the libwinpthread library using the command, but it is not able to detect that library:
`pacman -S mingw32/mingw-w64-i686-libwinpthread`

- Looks like there are other libraries which are missing, but I was not able to determine what is missing yet:
`-- Looking for explicit_memset
-- Looking for explicit_memset - not found
-- Looking for memset
-- Looking for memset - found
-- Looking for memset_s
-- Looking for memset_s - not found`

Is there another way to build on Windows? 
Thanks.

**Comment by AlexanderLanin** [comment]
Without any idea what I'm doing, so please forgive stupid mistakes:
* installed CMake
* installed MinGW + some C++ packages the installer offered
* start `C:\MinGW\msys\1.0\msys.bat`
* `git clone https://github.com/ccache/ccache.git`
* `mkdir build && cd build`
* `cmake ../ccache -G "MinGW Makefiles" -DZSTD_FROM_INTERNET=ON`
(To not generate visual studio sln file as we need to compile with gcc)
* `mingw32-make -j6` or `cmake --build .`

But I run into other problems like `_fileno` not available when using `--std=c++11`.
Maybe you can get some progress by providing the correct -G flag to cmake?

**Comment by AlexanderLanin** [comment]
Progress:

1.  Turns out fmt relies on GNU extensions for Windows. So the quick hack is to change `set(CMAKE_CXX_EXTENSIONS YES)` in CMakeLists.txt.
(run `rm -rf build/*`)
2. Next up is adding `-DWARNINGS_AS_ERRORS=Off` to the first cmake call. It's certainly not pretty but we get a bit further.
(run `rm -rf build/*`)
3. Start MinGW Installation Manager, select All Packages on the left and install mingw32-pthreads-w32-dev
(Restart msys2.bat and `rm -rf build/*`)

Current state:
\<thread\> does not contain `std::thread`

**Comment by Shionigami** [comment]
Thanks a lot for your hints :)

I did tests with my environment, using MinGW64. I also installed a lot of additional packages, and I am able to run the build with cmake (I completely forgot about the flag for MinGW...), and I got now issues with the WinUtils.cpp file. 
`In file included from D:\GitHub\ccache-master-test\src\Win32Util.cpp:19:
D:\GitHub\ccache-master-test\src\Win32Util.hpp:36:27: error: ‘DWORD’ was not declared in this scope
   36 | std::string error_message(DWORD error_code);
      |                           ^~~~~
D:\GitHub\ccache-master-test\src\Win32Util.cpp:37:1: error: redefinition of ‘std::string Win32Util::error_message’
   37 | error_message(DWORD error_code)
      | ^~~~~~~~~~~~~
In file included from D:\GitHub\ccache-master-test\src\Win32Util.cpp:19:
D:\GitHub\ccache-master-test\src\Win32Util.hpp:36:13: note: ‘std::string Win32Util::error_message’ previously declared here
   36 | std::string error_message(DWORD error_code);
      |             ^~~~~~~~~~~~~
D:\GitHub\ccache-master-test\src\Win32Util.cpp:37:15: error: ‘DWORD’ was not declared in this scope
   37 | error_message(DWORD error_code)
      |               ^~~~~
src\CMakeFiles\ccache_lib.dir\build.make:656: recipe for target 'src/CMakeFiles/ccache_lib.dir/Win32Util.cpp.obj' failed
mingw32-make.exe[2]: *** [src/CMakeFiles/ccache_lib.dir/Win32Util.cpp.obj] Error 1
CMakeFiles\Makefile2:447: recipe for target 'src/CMakeFiles/ccache_lib.dir/all' failed
mingw32-make.exe[1]: *** [src/CMakeFiles/ccache_lib.dir/all] Error 2
Makefile:181: recipe for target 'all' failed
mingw32-make.exe: *** [all] Error 2`

I'll take a look to it tomorrow, I got higher prio tasks at the moment. But we are getting closer to a successful build :D 

For the issue regarding the thread library, maybe you need to install the winpthread lib?

**Comment by Shionigami** [comment]
I added the following code to Win32Util.hpp:
`#ifdef __MINGW32__
#include <windows.h>
#endif`

It fixed the issue with DWORD, as expected, but then, I got, again, the issue with the winpthread library :(
`[ 81%] Linking CXX executable ccache.exe
/usr/lib/gcc/x86_64-pc-msys/9.3.0/../../../../x86_64-pc-msys/bin/ld: cannot find -lwinpthread
collect2: error: ld returned 1 exit status
CMakeFiles\ccache.dir\build.make:108: recipe for target 'ccache.exe' failed
mingw32-make.exe[2]: *** [ccache.exe] Error 1
CMakeFiles\Makefile2:282: recipe for target 'CMakeFiles/ccache.dir/all' failed
mingw32-make.exe[1]: *** [CMakeFiles/ccache.dir/all] Error 2
Makefile:181: recipe for target 'all' failed
mingw32-make.exe: *** [all] Error 2`

**Comment by jrosdahl** [comment]
Thanks @AlexanderLanin and @Shionigami for looking into this. This is uncharted terrain. :slightly_smiling_face:

**Comment by Shionigami** [comment]
Finally I was able to build ccache on Windows!
The main problem was that, even if I was using the MinGW64 provided with MSYS2, it was still using the MSYS compilers...

Here is how I did to build on Windows:
- Download MSYS2 from website: https://www.msys2.org/
- Run MinGW64: C:\msys64\mingw64.exe
- Install GCC package for MinGW64: `pacman -S mingw-w64-x86_64-gcc`
- Restart MinGW
- Install winpthread library: `pacman -S --needed mingw32/mingw-w64-i686-libwinpthread-git`
- Build with cmake:
`cmake .. -G "MinGW Makefiles" -DZSTD_FROM_INTERNET=ON `
`cmake --build .`

I am not 100% sure that there are all the necessary steps here, as I did a lot of tests and installation. I will try with a fresh installation later.

I also did some tests with a project running with GHS compiler, and it looks to work very well!
I also enabled the option base_dir, and changed the location of the build directory, but this did not work. I will investigate on that.
